src/coreclr/vm/codeman.cpp

```diff
@@ -3917,126 +3984,380 @@ StubCodeBlockKind EEJitManager::GetStubCodeBlockKind(RangeSection * pRangeSectio
+            // on ARM32 Thumb, the low bits indicate the thumb instruction set
+            pCode = ALIGN_DOWN(pCode, CODE_ALIGN);
+
+            size_t delta = pCode - pHp->mapBase;
+
+            size_t dwordIndex = GetDwordIndex(delta);
+            size_t nibbleIndex = GetNibbleIndex(delta);
+            DWORD value = ADDR2OFFS(delta);
+
+            DWORD mask  = POS2MASK(nibbleIndex);
+
+            value = value << Pos2ShiftCount(nibbleIndex);
+
+            PTR_DWORD pMap = pHp->pHdrMap;
+
+            // assert that we don't overwrite an existing offset
+            // the nibble is empty and the DWORD is not a pointer
+            _ASSERTE(!((*(pMap+dwordIndex)) & ~mask) && !IsPointer(*(pMap+dwordIndex)));
+
+            // It is important for this update to be atomic. Synchronization would be required with FindMethodCode otherwise.
+            *(pMap+dwordIndex) = ((*(pMap+dwordIndex)) & mask) | value;
+
+            size_t firstByteAfterMethod = delta + codeSize;
+            size_t nextDwordIndex = dwordIndex + 1;
+            while ((nextDwordIndex + 1) * BYTES_PER_DWORD <= firstByteAfterMethod)
+            {
+                // All of these DWORDs should be empty
+                _ASSERTE(!(*(pMap+nextDwordIndex)));
+                *(pMap+nextDwordIndex) = EncodePointer(delta);
+                nextDwordIndex++;
+            }
+        }
+
+        static void DeleteUnlocked(HeapList *pHp, TADDR pCode)
         {
-            tmp = tmp >> NIBBLE_SIZE;
-            startPos--;
+            CONTRACTL {
+                NOTHROW;
+                GC_NOTRIGGER;
+            } CONTRACTL_END;
+
+            _ASSERTE(pCode >= pHp->mapBase);
+
+            // remove bottom two bits to ensure alignment math
+            // on ARM32 Thumb, the low bits indicate the thumb instruction set
+            pCode = ALIGN_DOWN(pCode, CODE_ALIGN);
+
+            size_t delta = pCode - pHp->mapBase;
+
+            size_t dwordIndex = GetDwordIndex(delta);
+            size_t nibbleIndex = GetNibbleIndex(delta);
+
+            DWORD mask  = POS2MASK(nibbleIndex);
+
+            PTR_DWORD pMap = pHp->pHdrMap;
+
+            // assert that the nibble is not empty and the DWORD is not a pointer
+            _ASSERTE(((*(pMap+dwordIndex)) & ~mask) && !IsPointer(*(pMap+dwordIndex)));
+
+            // delete the relevant nibble
+            // It is important for this update to be atomic. Synchronization would be required with FindMethodCode otherwise.
+            *(pMap+dwordIndex) = ((*(pMap+dwordIndex)) & mask);
+
+            // the last DWORD of the nibble map is reserved to be empty for bounds checking
+            PTR_DWORD pNextDword = pMap + dwordIndex + 1;
+            while (IsPointer(*pNextDword) && DecodePointer(*pNextDword) == delta){
+                // The next DWORD is a pointer to the nibble being deleted, so we can delete it
+                *pNextDword = 0;
+                pNextDword++;
+            }
```

Comment:
might be important to note here that the write of the nibble and of the following DWORDS do not need to be synchronized with respect to each other because the read algorithm can deal with one or the other being out of order.